pytest_mh/utils/auditd.py

```diff
@@ -0,0 +1,133 @@
+        :type avc_filter: str | None, optional
+        """
+        super().__init__(host)
+
+        self.avc_mode: Literal["fail", "warn", "ignore"] = avc_mode
+        self.avc_filter: str | None = avc_filter
+
+        self.artifacts: set[str] = {"/var/log/audit/audit.log"}
+        self._backup: str | None = None
+
+    def setup(self) -> None:
+        """
+        Create backup of audit logs and clear them for current test run.
+        """
+        super().setup()
+
+        result = self.host.ssh.run(
+            """
+            set -e
+            tmp=`mktemp -d`
+            cp -r --archive /var/log/audit "$tmp"
+            truncate --size 0 /var/log/audit/audit.log*
+            echo $tmp
+            """,
+            log_level=SSHLog.Error,
+        )
+
+        self._backup = result.stdout.strip()
+
+    def teardown(self) -> None:
+        """
+        Restore previous audit logs from backup and remove the backup.
+        """
+        if self._backup is not None:
+            self.host.ssh.run(
+                f"""
+                set -e
+
+                for f in "{self._backup}"/audit/audit.log*; do
+                    name=`basename "$f"`
+                    cat "$f" > "/var/log/audit/$name"
+                done
+
+                rm -fr "{self._backup}"
+                """,
+                log_level=SSHLog.Error,
+            )
+
+        return super().teardown()
+
+    def pytest_report_teststatus(
+        self, report: pytest.CollectReport | pytest.TestReport, config: pytest.Config
+    ) -> tuple[str, str, str | tuple[str, dict[str, bool]]] | None:
+        """
+        Report AVC denial error if found and matches requested filter.
+
+        :param report: Pytest report
+        :type report: pytest.CollectReport | pytest.TestReport
+        :param config: Pytest config
+        :type config: pytest.Config
+        :return: Pytest test status
+        :rtype: tuple[str, str, str | tuple[str, dict[str, bool]]] | None
+        """
+        if report.when != "call":
+            return None
+
+        if self.avc_mode == "ignore":
+            return None
+
+        self.logger.info("Checking for AVC denials")
+
+        result = self.host.ssh.run(
+            "ausearch --input-logs -m AVC,USER_AVC", raise_on_error=False, log_level=SSHLog.Silent
+        )
+        if result.rc:
+            return None
+
+        records = result.stdout
+        if not records:
+            return None
+
+        error_msg = f"AVC denial was detected. Check audit log of {report.nodeid}."
+        if self.avc_filter:
+            match = re.search(self.avc_filter, records)
+            if match is None:
+                return None
+
+            error_msg = (
+                f"AVC denial was detected with regex filter '{self.avc_filter}'. Check audit log of {report.nodeid}."
+            )
+
+        match self.avc_mode:
+            case "fail":
```

Comment:
It would make more sense to set the test to error status instead.
We would like to distinguish here different states:
1. test itself passed but AVC happened - ERROR or WARNING. (AVC happened but functionality works as expected)
2. test did not pass and avc is present - FAIL. (This could be a fatal avc blocking part of functionality test testing )
We might want to ignore avcs altogether if there is a know issue and we want to deliver results.

Reply:
Thanks, this makes sense. I change the semantics to keep track of test outcome.

```

Selected tests will use the following hosts:
  audit: audit.test

collected 545 items / 544 deselected / 1 selected                                                                                                                                                                                            

tests/test_audit.py::test_audit (audit) PASSED/AVC DENIAL

======= 544 deselected, 1 AVC DENIALS in 0.94s =======
```

FAILED/PASSED is kept in the verbose output and all tests with AVC errors are counted towards "AVC DENIALS" at the end of the test. If "fail" mode is set, then non-zero return code is returned from pytest. If "warn" is set then test result is respected - so if test pass, zero exit code will be returned.